core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerde.java

```diff
@@ -165,161 +136,1043 @@ private static MarkerSet getSerializedPageMarkerSet(Slice serializedPage)
+                return;
+            }
+            // expand page output buffer
+            buffers[buffers.length - 1].ensureCapacityFor(bytes);
+
+            compress();
+            encrypt();
+        }
+
+        private void compress()
+        {
+            if (this.compressor.isEmpty()) {
+                return;
+            }
+            Compressor compressor = this.compressor.get();
+
+            WriteBuffer sourceBuffer = buffers[0];
+            WriteBuffer sinkBuffer = buffers[1];
+
+            int maxCompressedLength = compressor.maxCompressedLength(sourceBuffer.getPosition());
+            sinkBuffer.ensureCapacityFor(maxCompressedLength + Integer.BYTES);
+
+            int uncompressedSize = sourceBuffer.getPosition();
+            int compressedSize = compressor.compress(
+                    sourceBuffer.getSlice().byteArray(),
+                    sourceBuffer.getSlice().byteArrayOffset(),
+                    uncompressedSize,
+                    sinkBuffer.getSlice().byteArray(),
+                    sinkBuffer.getSlice().byteArrayOffset() + sinkBuffer.getPosition() + Integer.BYTES,
+                    maxCompressedLength);
+
+            boolean compressed = uncompressedSize * MINIMUM_COMPRESSION_RATIO > compressedSize;
+            int blockSize;
+            if (!compressed) {
+                System.arraycopy(
+                        sourceBuffer.getSlice().byteArray(),
+                        sourceBuffer.getSlice().byteArrayOffset(),
+                        sinkBuffer.getSlice().byteArray(),
+                        sinkBuffer.getSlice().byteArrayOffset() + sinkBuffer.getPosition() + Integer.BYTES,
+                        uncompressedSize);
+                blockSize = uncompressedSize;
+            }
+            else {
+                blockSize = compressedSize;
+            }
+
+            sinkBuffer.writeInt(createBlockMarker(compressed, blockSize));
+            sinkBuffer.skip(blockSize);
+
+            sourceBuffer.reset();
+        }
+
+        private static int createBlockMarker(boolean compressed, int size)
+        {
+            if (compressed) {
+                return size | COMPRESSED_BLOCK_MASK;
+            }
+            return size;
+        }
+
+        private void encrypt()
+        {
+            if (encryptionKey.isEmpty()) {
+                return;
+            }
+            Cipher cipher = initCipher(encryptionKey.get());
+            byte[] iv = cipher.getIV();
+
+            WriteBuffer sourceBuffer = buffers[buffers.length - 2];
+            WriteBuffer sinkBuffer = buffers[buffers.length - 1];
+
+            int maxEncryptedSize = cipher.getOutputSize(sourceBuffer.getPosition()) + iv.length;
+            sinkBuffer.ensureCapacityFor(maxEncryptedSize
+                    // to store encrypted block length
+                    + Integer.BYTES
+                    // to store initialization vector
+                    + iv.length);
```

Comment:
The IV block is actually written twice. Once in plain text explicitly (needed to initialize the Cipher for decryption) and once encrypted implicitly by the Cipher to perform decryption validation. The size of the encrypted IV written explicitly by the Cipher is accounted in `maxEncryptedSize` while the size of a plan text IV is not (as in theory it can be sent by an alternative communication channel).

Reply:
Gotcha, didn’t notice that in the logic while reviewing earlier. The extra IV for validation wouldn’t be necessary if we switched to AES/GCM and would also eagerly detect ciphertext corruption without a separate output checksumming scheme, but performance is I think just slightly worse than CBC last I checked.